File: app/src/domain/usage-range.ts
export type UsageRangePresetId = "today" | "week" | "month" | "all";

export interface UsageRangePreset {
  id: UsageRangePresetId;
  label: string;
  metaLabel: string;
  trailingDays: number | null;
  windowLabel: string;
}

export const DEFAULT_USAGE_RANGE_PRESET_ID: UsageRangePresetId = "week";

export const USAGE_RANGE_PRESETS: UsageRangePreset[] = [
  {
    id: "today",
    label: "Today",
    metaLabel: "1D",
    trailingDays: 1,
    windowLabel: "Today",
  },
  {
    id: "week",
    label: "Week",
    metaLabel: "7D",
    trailingDays: 7,
    windowLabel: "Last 7 days",
  },
  {
    id: "month",
    label: "Month",
    metaLabel: "30D",
    trailingDays: 30,
    windowLabel: "Last 30 days",
  },
  {
    id: "all",
    label: "All time",
    metaLabel: "\u221e",
    trailingDays: null,
    windowLabel: "All time",
  },
];

const RANGE_ORDER: Record<UsageRangePresetId, number> = {
  today: 0,
  week: 1,
  month: 2,
  all: 3,
};

export function getUsageRangePreset(range: UsageRangePresetId): UsageRangePreset {
  return USAGE_RANGE_PRESETS.find((preset) => preset.id === range) ?? USAGE_RANGE_PRESETS[1]!;
}

export function getUsageRangeOrder(range: UsageRangePresetId): number {
  return RANGE_ORDER[range];
}

export function usageRangeCovers(
  cachedRange: UsageRangePresetId,
  requestedRange: UsageRangePresetId,
): boolean {
  return getUsageRangeOrder(cachedRange) >= getUsageRangeOrder(requestedRange);
}

File: src-tauri/src/usage_range.rs
use chrono::{DateTime, Datelike, Local, TimeZone, Utc};

use crate::models::{UsageRangePresetId, UsageWindow};

pub const DEFAULT_USAGE_RANGE_PRESET_ID: UsageRangePresetId = UsageRangePresetId::Week;

pub struct UsageDateWindow {
  pub usage_window: UsageWindow,
}

struct UsageRangePreset {
  trailing_days: Option<i64>,
  window_label: &'static str,
}

fn get_usage_range_preset(range: UsageRangePresetId) -> UsageRangePreset {
  match range {
    UsageRangePresetId::Today => UsageRangePreset {
      trailing_days: Some(1),
      window_label: "Today",
    },
    UsageRangePresetId::Week => UsageRangePreset {
      trailing_days: Some(7),
      window_label: "Last 7 days",
    },
    UsageRangePresetId::Month => UsageRangePreset {
      trailing_days: Some(30),
      window_label: "Last 30 days",
    },
    UsageRangePresetId::All => UsageRangePreset {
      trailing_days: None,
      window_label: "All time",
    },
  }
}

fn start_of_local_day(value: DateTime<Local>) -> DateTime<Local> {
  Local
    .with_ymd_and_hms(value.year(), value.month(), value.day(), 0, 0, 0)
    .single()
    .expect("valid local day boundary")
}

pub fn create_usage_date_window(
  now: DateTime<Utc>,
  range: UsageRangePresetId,
  earliest_available_at: Option<DateTime<Utc>>,
) -> UsageDateWindow {
  let preset = get_usage_range_preset(range);
  let now_local = now.with_timezone(&Local);
  let until_local = start_of_local_day(now_local);
  let mut since_local = start_of_local_day(earliest_available_at.unwrap_or(now).with_timezone(&Local));

  if let Some(trailing_days) = preset.trailing_days {
    since_local = until_local - chrono::Duration::days(trailing_days - 1);
  }

  UsageDateWindow {
    usage_window: UsageWindow {
      label: preset.window_label.to_string(),
      since: since_local.with_timezone(&Utc).to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
      until: now.to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
    },
  }
}
